### PR TITLE
(#106) Skeleton should track method-method interactions

### PR DESCRIPTION
**Pull request by llorllale**
as per #106

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#153](https://github.com/yegor256/jpeek/issues/153) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#153](https://github.com/yegor256/jpeek/issues/153) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/main/java/org/jpeek/Skeleton.java [hunk 1]
```diff
@@ -344,6 +351,23 @@ public void visitFieldInsn(final int opcode,
                     }
                     Skeleton.Visitor.this.dirs.set(attr).up().up().up().up();
                 }
+
+                @Override
+                public void visitMethodInsn(final int opcode,
```
Comment (amihaiemil) [inline src/main/java/org/jpeek/Skeleton.java]: @llorllale what does ``insn`` mean? Can you leave a JavaDoc over this method?

File: src/main/java/org/jpeek/Skeleton.java [hunk 2]
```diff
@@ -332,7 +339,7 @@ public void visitFieldInsn(final int opcode,
                             "methods/method[@name='%s' and @desc='%s']",
                             mtd, desc
                         )
-                    ).strict(1).addIf("ops").add("op");
```
Comment (amihaiemil) [inline src/main/java/org/jpeek/Skeleton.java]: @llorllale why did you remove ``strict(1)``? :D

File: src/main/resources/org/jpeek/xsd/skeleton.xsd [hunk 3]
```diff
@@ -64,6 +64,15 @@ SOFTWARE.
                       </xs:sequence>
                     </xs:complexType>
                   </xs:element>
+                  <!--
+                   @todo #106:30min Refactor <ops>.
```
Comment (amihaiemil) [inline src/main/resources/org/jpeek/xsd/skeleton.xsd]: @llorllale Why does the ``<method_calls>`` element have to be inside ``<ops>`` -- what is ops, actually? To me it seems better if it is inside element ``<method>``, as you did it now.

File: src/main/java/org/jpeek/Skeleton.java (hunk 1 above)
Comment (llorllale) [reply src/main/java/org/jpeek/Skeleton.java]: `insn` stands for "instruction" - we're overriding [MethodVisitor.visitMethodInsn](http://asm.ow2.org/asm60/javadoc/user/org/objectweb/asm/MethodVisitor.html#visitMethodInsn-int-java.lang.String-java.lang.String-java.lang.String-boolean-).

Given `@Override` and this explanation, do you still feel a javadoc is needed?

File: src/main/resources/org/jpeek/xsd/skeleton.xsd (hunk 3 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/xsd/skeleton.xsd]: `method/ops` contains all operations of the method, which includes calls to methods.

**Comment by llorllale** [comment]
@amihaiemil I've added the `strict` checks - thanks for pointing that out

**Comment by amihaiemil** [comment]
@rultor merge

**Comment by amihaiemil** [comment]
@yegor256 ping

File: src/main/resources/org/jpeek/xsd/skeleton.xsd (hunk 3 above)
Comment (yegor256) [reply src/main/resources/org/jpeek/xsd/skeleton.xsd]: @llorllale I agree with @amihaiemil , this design looks weird. Let's go with this:

```xml
<ops>
  <op code="get">attr1</op>
  <op code="set">attr2</op>
  <op code="call">the_method_name</op> <!-- this is the new "code" we are introducing -->
</ops>
```

**Comment by yegor256** [comment]
@llorllale see above

File: src/main/resources/org/jpeek/xsd/skeleton.xsd (hunk 3 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/xsd/skeleton.xsd]: @yegor256 your proposal is missing the fqn of the class that the method belongs to. Where does it go?

**Comment by amihaiemil** [comment]
@llorllale How are we here, what's the status?

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@yegor256 here, for instance; we're still waiting for the DEV to do the last changes. What can I do to speed things up and not lose the task?

**Comment by llorllale** [comment]
@amihaiemil first of all, the policy says that the ten days rule does not apply to code reviews, so it looks like a bug in 0crat.

Second, this situation threw me in a loop because `REV` had already approved the PR but `ARC` overrode with his objection, so it took me a little more time to decide what to do. Also, I asked the `ARC` a question which is still unanswered.

The ticket reporter says he's fine with something like this which I don't like because it lacks structure:
```xml
<op code="call">java.lang.String.substring</op>
```

What I'm going to do in a couple of hours is try this instead:

```xml
<op code="call" owner="fully.qualified.name.of.owner.Class">method_name</op>
```

I'll try to see if not too many tests break, and will leave a puzzle or something saying that #156 needs to be fixed for field access.

WDYT?

**Comment by amihaiemil** [comment]
@llorllale 

> first of all, the policy says that the ten days rule does not apply to code reviews, so it looks like a bug in 0crat.

Now, it's not a bug now. It used to be a bug, but not anymore, it was introduced just now [here](https://github.com/zerocracy/farm/issues/537)

I don't know exactly, I did not follow the discussion. But if the ARCH or someone else does not answer to your question, just be shameless and ping them constantly, spam them until they answer. They will, eventually :D

**Comment by yegor256** [comment]
@llorllale if you don't know what to do or don't have time - do something and leave it as is, with a puzzle. Just don't way. Don't aim for perfection, aim for speed. That's the main rule.

**Comment by llorllale** [comment]
@amihaiemil please review. A few notes:
* left a puzzle in `MetricsTest` because adding that extra `<op>` breaks some tests
* made small, trivial fix in `SkeletonTest.createsXml` to make it work with the addition of the new `<op>`

**Comment by llorllale** [comment]
ping @amihaiemil

**Comment by amihaiemil** [comment]
@llorllale look ok

**Comment by amihaiemil** [comment]
@rultor merge

**Comment by llorllale** [comment]
@amihaiemil sorry, I had forgotten to remove the old `<method_calls>` element definition from `skeleton.xsd`. Please check

**Comment by amihaiemil** [comment]
@llorllale looks ok. And you don't need to add something new to the XSD regarding the extra <op>, right? (I'm not so familiar with XSD/XSLT, sorry if it's a noob question)

**Comment by llorllale** [comment]
@amihaiemil 
Not quite sure I understand:

> you don't need to add something new to the XSD regarding the extra

The extra what?

In any case, for this solution we went as per Yegor's suggestion which does not involve any changes to `skeleton.xsd`. Does this answer your question?

**Comment by amihaiemil** [comment]
@llorllale You said above:

> left a puzzle in MetricsTest because **adding that extra &lt;op&gt;** breaks some tests

But yes, it's clear now, since there aren't any changes to skeleton.xsd.

**Comment by amihaiemil** [comment]
@llorllale the initial comment did not render well, that's why it didn't make sense; I fixed it now :)

**Comment by llorllale** [comment]
@amihaiemil yes, Yegor's solution is to add an extra `<op>` to the generated XML (not the XSD). This breaks some tests, some math needs to be done for those, that's why I left the puzzle.

**Comment by amihaiemil** [comment]
@rultor merge

**Comment by amihaiemil** [comment]
@yegor256 Can we merge this now? I think until Monday 0crat will try to take it away from me :)

**Comment by llorllale** [comment]
@amihaiemil just solved some conflicts with an updated `master` branch, please check

**Comment by amihaiemil** [comment]
@yegor256 don't forget about this one. Clock is ticking

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@rultor merge

**Comment by llorllale** [comment]
@yegor256 @amihaiemil pdd failed because of a pdf. What do do here?

**Comment by amihaiemil** [comment]
@yegor256 can you try to merge again?

**Comment by yegor256** [comment]
@rultor merge

**Comment by llorllale** [comment]
@amihaiemil @yegor256 the build is failing on this file: `papers/henderson96.pdf`. As you can see, I did not touch this file in this PR. What can we do here?

**Comment by amihaiemil** [comment]
@llorllale I think you should open a bug about it. Probably someone merged a change in that file, in some other PR. Then, ask 0crat to add an impediment to your task so you don't lose it. However, I have no idea what to do about this PR, as far as I know, PRs cannot have impediments...

**Comment by llorllale** [comment]
@amihaiemil let's check now - I believe Yegor fixed the issue

**Comment by llorllale** [comment]
@amihaiemil @yegor256 opened #172 for the new build issue

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
Order was successfully finished: +15 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+525](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+9810](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#153](https://github.com/yegor256/jpeek/issues/153) is now out of scope
